src/server/socket.js

```diff
@@ -28,20 +44,112 @@ function SocketServer(simulatorProxy, project) {
     this._pendingEmits[SIM_HOST] = [];
     this._pendingEmits[DEBUG_HOST] = [];
 
-    /* Deferred are used to delay the actions until an event happens.
-     * Two major events are kept track of:
-     *      - When app-host connects (sends initial message)
-     *      - When sim-host notifies it's ready to serve an app-host.
-     */
-    this._whenAppHostConnected = Q.defer();
-    this._whenSimHostReady     = Q.defer();
+    // current client states
+    this._ahState = AH_STATE_DISCONNECTED;
+    this._shState = SH_STATE_DISCONNECTED;
 
     var config = this._simulatorProxy.config,
         telemetry = this._simulatorProxy.telemetry;
 
     this._liveReload = new LiveReload(project, telemetry, config.forcePrepare);
 }
 
+/**
+ * Handles protocol events triggered by the change of states in one of the
+ * clients (APP_HOST or SIM_HOST). Protocol events are those which depend on
+ * both APP_HOST and SIM_HOST.
+ */
+SocketServer.prototype._handleStateChange = function (appHostState, simHostState) {
+    var state = appHostState | simHostState;
+    switch (state) {
+        case AH_STATE_CONNECTED | SH_STATE_READY:
+            this._setupAppHostHandlers();
+            this._setupSimHostHandlers();
+            this._subscribeTo(APP_HOST, 'app-plugin-list', function (data) {
+                this._emitTo(SIM_HOST, 'app-plugin-list', data);
+                this._setAppHostState(AH_STATE_PLUGINS_READY);
+            }.bind(this), true);
+            this._emitTo(APP_HOST, 'init');
+            this._setAppHostState(AH_STATE_INIT_SENT);
+            break;
+        case AH_STATE_PLUGINS_READY | SH_STATE_READY:
+            this._subscribeTo(SIM_HOST, 'start', function () {
```

Comment:
> Does this introduce a new unintended synchronization point?

This synchronization was [there](https://github.com/Microsoft/cordova-simulate/pull/145/files#diff-c97dea31357a0880834e2407a2dcbc97L238) already. This one is just an explicit expression of it.

> Is it possible for sim-host to enter the READY state before app-host gets to the PLUGINS_READYstate, so we don't enter this case here, and yet sim-host emits a start event?

It does not work this way. `_handleStateChange` is called each time either of the clients changes the state. So which ones gets to which state first does not matter. The complexity of the previous implementation was exactly what you describe: the timing. In the new scheme of things, the timing of the state changes does not matter and each combination of the states of the clients is processed once.